client/go/outline/event/event.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2025 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package event provides a thread-safe mechanism for managing and triggering events.
+package event
+
+import (
+	"log/slog"
+	"slices"
+	"sync"
+
+	"github.com/Jigsaw-Code/outline-apps/client/go/outline/callback"
+)
+
+// EventName is a type alias for string that represents the name of an event.
+// Using a dedicated type improves type safety when working with event names.
+type EventName string
+
+var (
+	mu        sync.RWMutex
+	listeners = make(map[EventName][]callback.Token)
+)
+
+// AddListener adds a [callback.Callback] to a given [EventName].
+//
+// The provided callback will be called when the event is invoked, along with the event data.
+func AddListener(evt EventName, cb callback.Token) {
+	if evt == "" || cb == "" {
+		slog.Warn("empty event or callback are ignored")
+		return
+	}
+	mu.Lock()
+	defer mu.Unlock()
+
+	listeners[evt] = append(listeners[evt], cb)
+	slog.Debug("successfully subscribed to event", "event", evt, "callback", cb)
+}
+
+// RemoveListener removes a [callback.Callback] from the specified [EventName].
+//
+// Calling this function is safe even if the event has not been registered.
+func RemoveListener(evt EventName, cb callback.Token) {
```

Comment:
I worry about global events. It hurts composability and reusability, as it's impossible to track who is firing what events. Different components may conflict with the same events (e.g. "OnError").

We need a design that can compartmentalize events.

For example, you could have functions specific for the VPN events (`AddVPNAPIListener`). Or have a way to create independent "channels" (`AddListener(channel, event, callback)`). The channel would have a unique id.

Reply:
We also talked about having a `SetVPNListener`, which is easier to implement and reason about.

Reply:
I would not worry too much about "tracking who is firing the event" because the trigger object can include itself in the event data. (Also `OnError` is a bad event, events should be concrete, like `OnConnectionError`, `OnDisconnectionError`, etc.)

The event handler can then distinguish the trigger object by looking at the event details and then distribute it accordingly in TypeScript. That's also why I included the `VPNConn.ID` in the event, which uniquely represents the VPN connection object.

Reply:
I think it's ok, but the code safety and correctness relies on adhering to a strict structure that is not enforced by the compiler. A better approach would be to ensure we can create independent channels.

At a minimal create the object that represents the channel to stop using globals. This way we can have other methods that use their own channels.

Reply:
Same question as the one in `callback.go`, do you mean a `EventManager` and a `event.DefaultManager()` method?